[pytorch] Set alias analysis kind to FROM_SCHEMA for qadd, qmul, qclamp, qconcat

kevinbchen opened the pull request:
Summary:
Updated alias analysis kind to FROM_SCHEMA so input tensors can be marked as nonmutable
when appropriate, allowing for constant folding of these tensors.

Needed to update the schemas of the _out variants with annotations to mark the output input
tensor as aliased and mutable.

Test Plan:
```
import torch

class M(torch.nn.Module):
    def __init__(self):
        super(M, self).__init__()

    def forward(self, x):
        w = torch.tensor([3], dtype=torch.float)
        w = torch.quantize_per_tensor(w, 1.0, 0, torch.qint8)
        y = torch.tensor([3], dtype=torch.float)
        y = torch.quantize_per_tensor(w, 1.0, 0, torch.qint8)
        return torch.ops.quantized.add_out(x, w, y)

m = torch.jit.script(M())
torch._C._jit_pass_constant_propagation(m.graph)
print(m.graph)
```
```
graph(%self : __torch__.___torch_mangle_9.M,
      %x.1 : Tensor):
  %11 : int = prim::Constant[value=12]() # <ipython-input-11-1dd94c30cb58>:9:49
  %9 : float = prim::Constant[value=1.]() # <ipython-input-11-1dd94c30cb58>:9:41
  %10 : int = prim::Constant[value=0]() # <ipython-input-11-1dd94c30cb58>:9:46
  %36 : QInt8(1) = prim::Constant[value={3}]()
  %y.2 : Tensor = aten::quantize_per_tensor(%36, %9, %10, %11) # <ipython-input-11-1dd94c30cb58>:11:12
  %24 : Tensor = quantized::add_out(%x.1, %36, %y.2) # <ipython-input-11-1dd94c30cb58>:12:15
  return (%24)
```
As expected, the aten::quantize_per_tensor() for w is now folded. The aten::quantize_per_tensor()
for y is not folded, since that tensor is aliased/modified.

Differential Revision: D19910667

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qconcat.cpp:
```diff
@@ -110,20 +110,24 @@ static auto registry =
+            torch::RegisterOperators::options()
+                .aliasAnalysis(at::AliasAnalysisKind::FROM_SCHEMA)
+                .kernel<QCat<true>>(DispatchKey::QuantizedCPUTensorId))
         .op("quantized::cat_out(Tensor[] qx, int dim, Tensor out)"
```
this should be changed as well

kevinbchen replied:
Oops, thanks for catching, will fix.

jerryzh168 reviewed (approved):
Looks good. Thanks!

kevinbchen committed:
[pytorch] Set alias analysis kind to FROM_SCHEMA for qadd, qmul, qclamp, qconcat (#33359)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/33359

Updated alias analysis kind to FROM_SCHEMA so input tensors can be marked as nonmutable
when appropriate, allowing for constant folding of these tensors.

Needed to update the schemas of the _out variants with annotations to mark the output input
tensor as aliased and mutable.

Test Plan:
```
import torch

class M(torch.nn.Module):
    def __init__(self):
        super(M, self).__init__()

    def forward(self, x):
        w = torch.tensor([3], dtype=torch.float)
        w = torch.quantize_per_tensor(w, 1.0, 0, torch.qint8)
        y = torch.tensor([3], dtype=torch.float)
        y = torch.quantize_per_tensor(w, 1.0, 0, torch.qint8)
        return torch.ops.quantized.add_out(x, w, y)

m = torch.jit.script(M())
torch._C._jit_pass_constant_propagation(m.graph)
print(m.graph)
```
```
graph(%self : __torch__.___torch_mangle_9.M,
      %x.1 : Tensor):
  %11 : int = prim::Constant[value=12]() # <ipython-input-11-1dd94c30cb58>:9:49
  %9 : float = prim::Constant[value=1.]() # <ipython-input-11-1dd94c30cb58>:9:41
  %10 : int = prim::Constant[value=0]() # <ipython-input-11-1dd94c30cb58>:9:46
  %36 : QInt8(1) = prim::Constant[value={3}]()
  %y.2 : Tensor = aten::quantize_per_tensor(%36, %9, %10, %11) # <ipython-input-11-1dd94c30cb58>:11:12
  %24 : Tensor = quantized::add_out(%x.1, %36, %y.2) # <ipython-input-11-1dd94c30cb58>:12:15
  return (%24)
```
As expected, the aten::quantize_per_tensor() for w is now folded. The aten::quantize_per_tensor()
for y is not folded, since that tensor is aliased/modified.

Differential Revision: D19910667

fbshipit-source-id: 1c9d0d1df74d07008360478229c31c1fcd0c791f
